### ISSUE TITLE
stable rustc --version hangs forever

### ISSUE THREAD

**Issue by fuchsnj** [rust-lang/rust#56736]
starting with `nightly-2018-11-04` and anything later, just checking the version or doing anything with rustc causes the process to hang forever. It seems like it is waiting on a lock (no CPU usage)

Stable rust works fine, and well as any nightly version before this one.

OS is `Ubuntu 18.04.1 LTS`
Rust stable/nightly versions were installed with rustup

```
nathan@nathan-Precision-7510:~$ rustc +nightly-2018-11-04 --version
^C
nathan@nathan-Precision-7510:~$ rustc +nightly-2018-11-03 --version
rustc 1.32.0-nightly (8b096314a 2018-11-02)
```

The last few lines from `strace rustc +nightly-2018-11-04 --version` are
```
...
rt_sigprocmask(SIG_UNBLOCK, [RTMIN RT_1], NULL, 8) = 0
prlimit64(0, RLIMIT_STACK, NULL, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
readlink("/etc/malloc.conf", 0x7ffd8dd4a900, 4096) = -1 ENOENT (No such file or directory)
open("/proc/sys/vm/overcommit_memory", O_RDONLY) = 3
futex(0x7fbdfd2510c8, FUTEX_WAKE_PRIVATE, 2147483647) = 0
futex(0x7fbdfe014228, FUTEX_WAIT_PRIVATE, 2, NULL
```
It hangs on the `FUTEX_WAIT` call forever

**Comment by ehuss** [comment]
Another user reported the same issue here: https://github.com/rust-lang/cargo/issues/6384

That is the first release that removed jemalloc.  I would suspect that is related, but I don't have any ideas on how to reproduce.

**Comment by fuchsnj** [comment]
Here is a stacktrace of the stuck process
https://gist.github.com/fuchsnj/5612923e3613a915b65aece0dd920149

**Comment by fuchsnj** [comment]
Yes, I have ESET antivirus installed.

**Comment by alexcrichton** [comment]
cc @gnzlbg, do you know if jemalloc has a fix for this perhaps?

**Comment by ehuss** [comment]
It appears to be an issue with ESET, jemalloc 5, and rustc being built for an old kernel.

It looks like jemalloc 5 has started to use CLOEXEC. Since rust is built with a very old linux kernel, it has to use fcntl ([here](https://github.com/jemalloc/jemalloc/blob/61efbda7098de6fe64c362d309824864308c36d4/src/pages.c#L447)) instead of just passing O_CLOEXEC (which requires 2.6.23).  fcntl is intercepted by ESET which attempts to find the `open` symbol with dlsym.  dlsym requires calling calloc, which hangs since jemalloc is in the middle of initializing.

I have confirmed building rustc locally (with jemalloc) that it does not hang, presumably because it is using O_CLOEXEC.

I don't offhand see any workarounds (other than using a newer kernel).

**Comment by gnzlbg** [comment]
> I don't offhand see any workarounds (other than using a newer kernel).

The PR that started using CLOEXEC was https://github.com/jemalloc/jemalloc/pull/872 which fixed https://github.com/jemalloc/jemalloc/issues/528 . We could patch jemalloc to not use CLOEXEC when built for Rust, but... it looks to me that jemalloc is doing the right thing here, and that this is a corner case that should be handled in ESETs side. 

We should open a bug with ESET about their jemalloc 5 / fcntl / Rust support, maybe they can roll a fix quickly. Depending on their timeline, patching jemalloc to not use CLOEXEC shouldn't be hard.

**Comment by alexcrichton** [comment]
Opening a bug (if we can) with ESET sounds good to me for now, but if that doesn't pan out we can probably work around this and just not use cloexec there as it's a short-lived fd anyway

**Comment by gnzlbg** [comment]
> the rustc built via CI opts back into having jemalloc linked to rustc.

IIUC the _intent_ was for `rustc` to always depend on `jemalloc` by default, since that was the status-quo before that change, but to allow people to build it without `jemalloc`, e.g., if they want to use it in a system where `jemalloc` is not available.  It might be that this did not fully materialize.

**Comment by pnkfelix** [comment]
Yes, I too thought that was the *intent*. But...:

https://github.com/rust-lang/rust/blob/f4b07e0713b2d82417968db08cd0575734cdac0d/config.toml.example#L399-L401

----

It could very well be that *I* am wrong about my expectations, and that if one wants to replicate the CI build product, one should take care to actually run `configure` with args taken from e.g.:

https://github.com/rust-lang/rust/blob/f4b07e0713b2d82417968db08cd0575734cdac0d/.travis.yml#L33

(or with configure args taken from https://github.com/rust-lang/rust/blob/master/appveyor.yml, as appropriate to one's platform)

**Comment by pnkfelix** [comment]
I'm just going to open a separate issue about this discrepancy between the CI vs local builds, rather than continuing to clutter up this issue's comment thread. Sorry for the noise!

**Comment by pnkfelix** [comment]
T-compiler triage. This issue is tagged as a regression but no T-label, so no team has default responsibility for it. Based on the comments in the issue here, I do not think T-compiler is in a position to fix this; it seems to be probably a T-infra problem?

**Comment by fuchsnj** [comment]
As of Rust 1.32 (Jan 17th) this now affects the latest Rust stable version.

**Comment by aidanhs** [comment]
We discussed this in the infra team meeting a few weeks, and basically decided (given that this appears a strange interaction between jemalloc and ESET we're stuck in the middle of) to wait until beta (and stable!) to see if more people reported the issue.

Given that we've not seen more reports, unfortunately this isn't going to be something we prioritise - our hope is that either jemalloc or ESET fixes things.

That said - @turboladen, we're interested in your report. Did you manage to track anything down, e.g. via strace?

**Comment by turboladen** [comment]
@aidanhs unfortunately we didn't really get time to get much info on it. We ended up cloning over the AWS image we had for our same app from a different environment (beta was having the problem described in this ticket, staging was not); I do believe, however, that we had jemalloc installed on the beta instance (trying to help speed up Rails), but did _not_ have jemalloc on the staging instance.

**Comment by antekone** [comment]
Just FYI, ESET has fixed the issue in version `4.5.14.0`. So if anyone uses older product version and suffers from this hang, please try updating to `4.5.14.0` or later.
